js/ui/scheduler/utils.recurrence.js

```diff
@@ -1,854 +1,413 @@
 
-    return intervalObject;
-};
+        const duration = endDateUtc ? endDateUtc.getTime() - startDateUtc.getTime() : 0;
 
-const getDatesByRecurrenceException = function(ruleValues, date) {
-    const result = [];
+        this._initializeRRule(options, startDateUtc);
 
-    for(let i = 0, len = ruleValues.length; i < len; i++) {
-        result[i] = getDateByAsciiString(ruleValues[i], date);
-    }
+        const minTime = minDateUtc.getTime();
+        const leftBorder = this._getLeftBorder(options, minDateUtc, duration);
+        this.rRuleSet.between(leftBorder, maxDateUtc, true).forEach(date => {
+            const endAppointmentTime = date.getTime() + duration;
 
-    return result;
-};
-
-const dateIsRecurrenceException = function(date, recurrenceException) {
-    let result = false;
+            if(endAppointmentTime >= minTime) {
+                this._correctTimezoneOffset(date);
+                result.push(date);
+            }
+        });
 
-    if(!recurrenceException) {
         return result;
     }
 
-    const splitDates = recurrenceException.split(',');
-    const exceptDates = getDatesByRecurrenceException(splitDates, date);
-    const shortFormat = /\d{8}$/;
+    _initializeRRule(options, startDateUtc) {
+        const ruleOptions = RRule.parseString(options.rule);
+        const firstDayOfWeek = options.firstDayOfWeek;
 
-    for(let i = 0, len = exceptDates.length; i < len; i++) {
-        if(splitDates[i].match(shortFormat)) {
-            const diffs = getDatePartDiffs(date, exceptDates[i]);
+        ruleOptions.dtstart = startDateUtc;
 
-            if(diffs.years === 0 && diffs.months === 0 && diffs.days === 0) {
-                result = true;
-            }
-        } else {
-            if(date.getTime() === exceptDates[i].getTime()) {
-                result = true;
-            }
+        if(!ruleOptions.wkst && firstDayOfWeek) {
+            const weekDayNumbers = [6, 0, 1, 2, 3, 4, 5];
+            ruleOptions.wkst = weekDayNumbers[firstDayOfWeek];
         }
-    }
-
-    return result;
-};
-
-const doNextIteration = function(date, startIntervalDate, endIntervalDate, recurrenceRule, iterationCount) {
-    let matchCountIsCorrect = true;
 
-    endIntervalDate = endIntervalDate.getTime();
+        this._createRRule(ruleOptions);
 
-    if(recurrenceRule.until) {
-        if(recurrenceRule.until.getTime() < endIntervalDate) {
-            endIntervalDate = recurrenceRule.until.getTime();
-        }
-    }
-
-    if(recurrenceRule.count) {
-        if(iterationCount === recurrenceRule.count) {
-            matchCountIsCorrect = false;
+        if(options.exception) {
+            const splitDates = options.exception.split(',');
+            const exceptDates = this._getDatesByRecurrenceException(splitDates, startDateUtc);
+            for(let i = 0; i < exceptDates.length; i++) {
+                this.rRuleSet.exdate(this._getRRuleUtcDate(exceptDates[i]));
+            }
```

Comment:
Maybe it is better to use forEach?

Reply:
Very strange - I have already corrected this piece of code.
Sure, I've correct it